core/src/main/java/org/apache/iceberg/IncrementalDataTableScan.java

```diff
@@ -0,0 +1,124 @@
+    Preconditions.checkArgument(
+        table().snapshot(newToSnapshotId) != null, "toSnapshotId: %s does not exist", newToSnapshotId);
+    return new IncrementalDataTableScan(
+        tableOps(), table(), schema(), filter(), isCaseSensitive(), colStats(), selectedColumns(), options(),
+        newFromSnapshotId, newToSnapshotId);
+  }
+
+  @Override
+  public TableScan appendsAfter(Long newFromSnapshotId) {
+    final Snapshot currentSnapshot = table().currentSnapshot();
+    Preconditions.checkState(currentSnapshot != null, "Current snapshot is null");
+    return appendsBetween(newFromSnapshotId, currentSnapshot.snapshotId());
+  }
+
+  @Override
+  public CloseableIterable<FileScanTask> planFiles() {
+    //TODO publish an incremental appends scan event
+    List<Snapshot> snapshots = snapshotsWithin(table(), fromSnapshotId, toSnapshotId);
+    Iterable<CloseableIterable<FileScanTask>> files = Iterables.transform(snapshots, this::planFiles);
+    return CloseableIterable.concat(files);
+  }
+
+  private CloseableIterable<FileScanTask> planFiles(Snapshot snapshot) {
+    Predicate<ManifestFile> matchingManifests = manifest -> manifest.snapshotId() == snapshot.snapshotId();
+
+    Predicate<ManifestEntry> matchingManifestEntries =
+        manifestEntry ->
+            manifestEntry.snapshotId() == snapshot.snapshotId() &&
+            manifestEntry.status() == ManifestEntry.Status.ADDED;
+
+    return planFiles(
+        tableOps(), snapshot, filter(), isCaseSensitive(), colStats(), matchingManifests, matchingManifestEntries);
+  }
+
+  @Override
+  @SuppressWarnings("checkstyle:HiddenField")
+  protected TableScan newRefinedScan(
+          TableOperations ops, Table table, Long snapshotId, Schema schema, Expression rowFilter,
+          boolean caseSensitive, boolean colStats, Collection<String> selectedColumns,
+          ImmutableMap<String, String> options) {
+    return new IncrementalDataTableScan(
+            ops, table, schema, rowFilter, caseSensitive, colStats, selectedColumns, options,
+            fromSnapshotId, toSnapshotId);
+  }
+
+  private static List<Snapshot> snapshotsWithin(Table table, long fromSnapshotId, long toSnapshotId) {
+    List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+    int toSnapshotIdIndex = snapshotIds.indexOf(toSnapshotId);
+    Preconditions.checkState(toSnapshotIdIndex != -1, "Undefined state: toSnapshotId: %s not found", toSnapshotId);
+    int fromSnapshotIdIndex = snapshotIds.indexOf(fromSnapshotId);
+    Preconditions.checkState(
+        fromSnapshotIdIndex != -1, "Undefined state: fromSnapshotId: %s not found", fromSnapshotId);
+
+    List<Snapshot> snapshots = Lists.newArrayList();
+    for (int i = toSnapshotIdIndex; i < fromSnapshotIdIndex; i++) {
+      Snapshot snapshot = table.snapshot(snapshotIds.get(i));
+      // for now, incremental scan supports only appends
+      if (snapshot.operation().equals(DataOperations.APPEND)) {
+        snapshots.add(snapshot);
+      }
```

Comment:
I think this needs an `else` case that validates we aren't skipping over `overwrite` commits. I think it is okay to skip `rewrite` and `delete`, but `overwrite` adds data so I'm not sure about skipping it. I could be convinced to add a config option to do this later, but it seems dangerous to me to skip it automatically. We can always allow more later, but it's hard to disallow something that always worked.